File: 227-basic-calculator-ii/basic-calculator-ii.java
class Solution {
    public int calculate(String s) {
        Stack<Integer> stack = new Stack<>();
        s = s.replaceAll("\\s+", "");
        int p = 0;
        int num = 0;
        char sign = '+';
        while (p < s.length()) {
            if (s.charAt(p) >= '0' && s.charAt(p) <= '9') {
                num = num * 10 + s.charAt(p) - '0';
            } 
            // find the last sign before this character
            if (!Character.isDigit(s.charAt(p)) || p == s.length() - 1) {
                if (sign == '+') {
                    stack.add(num);
                } else if (sign == '-') {
                    stack.add(-num);
                } else if (sign == '*') {
                    stack.add(stack.pop()*num);
                } else if(sign == '/'){
                    stack.add(stack.pop()/num);
                }
                sign = s.charAt(p);
                num = 0;
            }
            p++;
        }
        int res = 0;
        while (!stack.isEmpty()) {
            res += stack.pop();
        }
        return res;
    }
}

File: 227-basic-calculator-ii/basic-calculator-ii.py
class Solution:
    def calculate(self, s: str) -> int:
        s = s + "+"
        sign = "+"
        cur = 0
        prev = 0
        res = 0
        for c in s:
            # c is digit
            if c.isnumeric():
                # convert the digit to the total num
                cur = cur * 10 + int(c)
            else:
                if c != ' ':
                    if sign == "+":
                        res += prev
                        prev = cur
                    elif sign == "-":
                        res += prev
                        prev = -cur
                    elif sign == "*":
                        prev = prev * cur
                    elif sign == "/":
                        prev = math.trunc(prev/cur)
                    cur = 0
                    sign = c
        res += prev
        return res